sbom/spdx/encoder.go

```diff
@@ -0,0 +1,444 @@
+				if err != nil {
+					return nil, err
+				}
+
+				srcPkg, ok := pkgs[rSrcPkgId]
+				// Record the source package if we haven't seen it yet.
+				if !ok {
+					srcPkg = newSpdxPackageFromPackage(r.Package.Source)
+					srcPkg.PrimaryPackagePurpose = "SOURCE"
+					pkgs[rSrcPkgId] = srcPkg
+					pkgIds = append(pkgIds, rSrcPkgId)
+				}
+
+				rel := &v2_3.Relationship{
+					RefA:         v2common.MakeDocElementID("", string(pkg.PackageSPDXIdentifier)),
+					RefB:         v2common.MakeDocElementID("", string(srcPkg.PackageSPDXIdentifier)),
+					Relationship: "GENERATED_FROM",
+				}
+				pkgRels[rPkgId] = append(pkgRels[rPkgId], rel)
+			}
+		} else if pkg.PrimaryPackagePurpose == "SOURCE" {
+			// If we recorded a source package when we found it as an r.Package.Source,
+			// we need to record any missing information we didn't know about previously.
+			pkg.FilesAnalyzed = true
+			if pkg.PackageFileName == "" {
+				pkgDB := ""
+				for _, env := range ir.Environments[r.Package.ID] {
+					if env.PackageDB != "" {
+						pkgDB = env.PackageDB
+						break
+					}
+				}
+				pkg.PackageFileName = pkgDB
+			}
+		}
+
+		// Record Distributions for this package.
+		if r.Distribution != nil {
+			rDistId, err := strconv.Atoi(r.Distribution.ID)
+			if err != nil {
+				return nil, err
+			}
+
+			dist, ok := dists[rDistId]
+			// Record the Distribution if we haven't seen it yet.
+			if !ok {
+				dist = newSpdxPackageFromDistribution(r.Distribution)
+				dists[rDistId] = dist
+				distIds = append(distIds, rDistId)
+			}
+
+			rel := &v2_3.Relationship{
+				RefA:         v2common.MakeDocElementID("", string(pkg.PackageSPDXIdentifier)),
+				RefB:         v2common.MakeDocElementID("", string(dist.PackageSPDXIdentifier)),
+				Relationship: "CONTAINED_BY",
+			}
+			pkgRels[rPkgId] = append(pkgRels[rPkgId], rel)
+		}
+
+		// Record Repositories for this package.
+		if r.Repository != nil {
+			rRepoId, err := strconv.Atoi(r.Repository.ID)
+			if err != nil {
+				return nil, err
+			}
+
+			repo, ok := repos[rRepoId]
+			// Record the Repository if we haven't seen it yet.
+			if !ok {
+				repo = newSpdxPackageFromRepository(r.Repository)
+
+				repos[rRepoId] = repo
+				repoIds = append(repoIds, rRepoId)
+			}
+
+			rel := &v2_3.Relationship{
+				RefA:         v2common.MakeDocElementID("", string(pkg.PackageSPDXIdentifier)),
+				RefB:         v2common.MakeDocElementID("", string(repo.PackageSPDXIdentifier)),
+				Relationship: "CONTAINED_BY",
+			}
+			pkgRels[rPkgId] = append(pkgRels[rPkgId], rel)
+		}
+	}
+
+	// Now that we have all the data necessary to create the SPDX document,
+	// we need to order it since the IndexRecords aren't in a deterministic order.
+	// This is particular helpful for testing, but it wouldn't be unreasonable
+	// for a user to want to diff different versions of an SPDX of the same IndexReport.
+	slices.Sort(pkgIds)
+	for _, id := range pkgIds {
+		if ctx.Err() != nil {
```

Comment:
I actually have no idea if this is worth doing or not. Fine to keep as-is, but I guess I still don't really know every situation when we should check the context's status.

Google's AI overview says to do it upon each iteration of a loop like this, though take that with a grain of salt... Guess it's not hurting anyone here

Reply:
Yeah, I wasn't sure either. I figured if we have the context, we might as well? I'm fine to throw this out, though.